### DISCUSSION TITLE
`Entity::insert` doesn't call `ActiveModelBehaviour::before_save`

### DISCUSSION THREAD

**Discussion by teenjuna** [SeaQL/sea-orm#1532] [Ideas]
## Description

The title describes it all. I didn't check, but it's possible that other associated methods also don't call these hooks.

## Steps to Reproduce

Print a log message inside `before_save` hook and do `entity::model::Entity::insert(model).exec(conn)`

### Expected Behavior

Hooks must be called.

### Reproduces How Often

Always.

**Comment by billy1624** [comment]
This issue might also applies to `update` & `delete`

**Comment by Diwakar-Gupta** [comment]
Hook's are called from `ActiveModel.{insert, update, delete}` only and not from
`Entity:: { insert, insert_many, update, update_many, delete, delete_many, delete_by_id }`.

Here `ActiveModel.<method>` call's corresponding `Entity::<method>` 
https://github.com/SeaQL/sea-orm/blob/47c1a9b145080f66fbe0b91c286510487b3096e9/src/entity/active_model.rs#L292-L303
https://github.com/SeaQL/sea-orm/blob/47c1a9b145080f66fbe0b91c286510487b3096e9/src/entity/base_entity.rs#L365-L370

I think we can add new methods to `ActiveModel:: { insert_many, update_many, delete_many } ` which will call all the hooks for all model's then call corresponding methods in `Entity::<method>`. What do you think @billy1624 ?

After that we can change the docs to make user utilize methods from `ActiveModel`;

**Comment by billy1624** [comment]
I think it's expected that the `ActiveModelBehaviour` hooks will only be run when you invoke the `insert`, `update` or `delete` in `ActiveModelTrait`.

**Comment by billy1624** [comment]
Ping @tyt2y3

**Comment by Diwakar-Gupta** [comment]
> I think it's expected that the `ActiveModelBehaviour` hooks will only be run when you invoke the `insert`, `update` or `delete` in `ActiveModelTrait`.

Agree, but when multiple models were updated using `Entity::update` hooks are not called.

Currently `ActiveModelTrait` can be used to update single model at a time for that it works as expected.

The idea is to add a method in  `ActiveModelTrait` which can update multiple models from list and call hook for all of them.

**Comment by tyt2y3** [comment]
I think it is what the API design imply. https://www.sea-ql.org/SeaORM/docs/internal-design/architecture/

`ActiveModelBehaviour`, as it's name suggest, is attached to an ActiveModel.

While the basic form of `Entity::insert` does the same thing as `ActiveModel::save`, in other cases, such as:

```rust
fruit::Entity::update(orange.clone()).filter(fruit::Column::Name.contains("orange")).exec(&db)
```

By the time we are about to execute the statement, the query could have been changed, and thus we cannot accurately handle `before_save`.

> when multiple models were updated using Entity::update hooks are not called

After an update many statement, we don't have all the updated models in memory. We will need to fetch all the affected models, then invoke the `after_save` one by one. This is an expensive operation, so we cannot do it inside `exec`. May be, an `exec_and_return_all_models` (this might be a bad name).

Moving this to discussion and see if we can work out a proposal.
